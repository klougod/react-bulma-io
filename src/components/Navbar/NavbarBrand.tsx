import React, { Children, ReactElement, CSSProperties, MouseEvent } from 'react'
import PropTypes from 'prop-types'

export const NavbarBrand = ({ className, menuRef, children }: any) => {
  const handleBurgerClick = (e: MouseEvent<HTMLButtonElement>) => {
    const target = e.target as HTMLTextAreaElement
    if (target.classList.contains('is-active')) {
      target.classList.remove('is-active')
      menuRef.current.classList.remove('is-active')
    } else {
      target.classList.add('is-active')
      menuRef.current.classList.add('is-active')
    }
  }

  const burgerStyle: CSSProperties = {
    backgroundColor: 'transparent',
    border: 0
  }

  const noPointerEvents: CSSProperties = {
    pointerEvents: 'none'
  }

  const NavBurger = () => (
    <button
      onClick={handleBurgerClick}
      className='navbar-burger'
      style={burgerStyle}
      aria-label='abrir menu'
      aria-expanded='false'
    >
      <span aria-hidden='true' style={noPointerEvents} />
      <span aria-hidden='true' style={noPointerEvents} />
      <span aria-hidden='true' style={noPointerEvents} />
    </button>
  )

  return (
    <div className={`navbar-brand ${className}`}>
      {Children.toArray(children).map((item: ReactElement, index) => {
        return React.cloneElement(item, {
          key: index,
          className: `${item.props.className} navbar-item`
        })
      })}
      <NavBurger />
    </div>
  )
}

NavbarBrand.defaultProps = {
  className: '',
  menuRef: null
}

NavbarBrand.propTypes = {
  className: PropTypes.string,
  menuRef: PropTypes.any,
  children: PropTypes.node
}

export default NavbarBrand
