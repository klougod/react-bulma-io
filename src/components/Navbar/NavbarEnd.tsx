import React, { Children, ReactElement } from 'react'
import PropTypes from 'prop-types'

export const NavbarEnd = ({ className, children }: any) => (
  <div className={`navbar-end ${className}`}>
    {Children.toArray(children).map((item: ReactElement, index) =>
      React.cloneElement(item, {
        key: index,
        className: `${item.props.className} navbar-item`
      })
    )}
  </div>
)

NavbarEnd.defaultProps = {
  className: ''
}

NavbarEnd.propTypes = {
  className: PropTypes.string,
  children: PropTypes.node
}

export default NavbarEnd
