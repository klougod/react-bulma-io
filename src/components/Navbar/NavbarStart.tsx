import React, { Children, ReactElement } from 'react'
import PropTypes from 'prop-types'

export const NavbarStart = ({ className, children }: any) => (
  <div className={`navbar-start ${className}`}>
    {Children.toArray(children).map((item: ReactElement, index) =>
      React.cloneElement(item, {
        key: index,
        className: `${item.props.className} navbar-item`
      })
    )}
  </div>
)

NavbarStart.defaultProps = {
  className: ''
}

NavbarStart.propTypes = {
  className: PropTypes.string,
  children: PropTypes.node
}

export default NavbarStart
