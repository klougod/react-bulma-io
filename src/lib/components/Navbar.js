import React, { useRef, Children } from 'react'
import PropTypes from 'prop-types'

export const Navbar = props => {

  const {
    className,
    menuClass,
    children
  } = props

  const menuRef = useRef(null)
  const childrenArray = Children.toArray(children)

  return (
    <nav className={`navbar ${className}`} role="navigation" aria-label="main navigation">
      { childrenArray.filter(c => c.type.name === "NavbarBrand") }
      <div ref={menuRef} className={`navbar-menu ${menuClass}`}>
        { childrenArray.filter(c => c.type.name === "NavbarStart") }
        { childrenArray.filter(c => c.type.name === "NavbarEnd") }
      </div>
    </nav>
  )
}

Navbar.defaultProps = {
  className: '',
  menuClass: ''
}

Navbar.propTypes = {
  className: PropTypes.string,
  menuClass: PropTypes.string,
  children: PropTypes.node.isRequired
}

export default Navbar