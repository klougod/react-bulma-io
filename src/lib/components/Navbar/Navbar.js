import React, { useRef, Children } from 'react'
import PropTypes from 'prop-types'

export const Navbar = ({ className, menuClass, children }) => {
  const menuRef = useRef(null)
  const childrenArray = Children.toArray(children)

  return (
    <nav className={`navbar ${className}`} role="navigation" aria-label="main navigation">
      {
        childrenArray.map(c => c.type.name === "NavbarBrand" && React.cloneElement(c, { menuRef }))
      }
      <div ref={menuRef} className={`navbar-menu ${menuClass}`}>
        { childrenArray.filter(c => c.type.name === "NavbarStart") }
        { childrenArray.filter(c => c.type.name === "NavbarEnd") }
      </div>
    </nav>
  )
}

Navbar.defaultProps = {
  className: '',
  menuClass: ''
}

Navbar.propTypes = {
  className: PropTypes.string,
  menuClass: PropTypes.string,
  children: PropTypes.node.isRequired
}

export default Navbar