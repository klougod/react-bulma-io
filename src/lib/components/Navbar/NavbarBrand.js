import React, { Children } from 'react'
import PropTypes from 'prop-types'

export const NavbarBrand = ({ className, menuRef, children }) => {
  const handleBurgerClick = e => {
    if (e.target.classList.contains("is-active")) {
      e.target.classList.remove("is-active");
      menuRef.current.classList.remove("is-active");
    } else {
      e.target.classList.add("is-active");
      menuRef.current.classList.add("is-active");
    }
  }

  const burgerStyle = {
    backgroundColor: 'transparent',
    border: 0
  }

  const noPointerEvents = {
    pointerEvents: 'none'
  }

  const NavBurger = () => (
    <button onClick={handleBurgerClick} className="navbar-burger" style={burgerStyle} aria-label="abrir menu" aria-expanded="false">
      <span aria-hidden="true" style={noPointerEvents} />
      <span aria-hidden="true" style={noPointerEvents} />
      <span aria-hidden="true" style={noPointerEvents} />
    </button>
  )

  return (
    <div className={`navbar-brand ${className}`}>
      {Children.toArray(children).map((item, index) => {
        return React.cloneElement(item, {
          key: index,
          className: `${item.props.className} navbar-item`
        })
      })}
      <NavBurger />
    </div>
  )
}

NavbarBrand.defaultProps = {
  className: '',
  menuRef: null
}

NavbarBrand.propTypes = {
  className: PropTypes.string,
  menuRef: PropTypes.any,
  children: PropTypes.node.isRequired
}

export default NavbarBrand