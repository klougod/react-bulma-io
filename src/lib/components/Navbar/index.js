export { default as Navbar } from './Navbar'
export { default as NavbarBrand } from './NavbarBrand'
export { default as NavbarStart } from './NavbarStart'
export { default as NavbarEnd } from './NavbarEnd'
