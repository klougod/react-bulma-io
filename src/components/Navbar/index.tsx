import React, { useRef, Children, ReactElement } from 'react'
import PropTypes from 'prop-types'
import NavbarStart from './NavbarStart'
import NavbarEnd from './NavbarEnd'
import NavbarBrand from './NavbarBrand'

const navbarBrandType = (<NavbarBrand />).type
const navbarStartType = (<NavbarStart />).type
const navbarEndType = (<NavbarEnd />).type

const Navbar = ({ className, menuClass, children }: any) => {
  const menuRef = useRef(null)
  const childrenArray = Children.toArray(children)

  return (
    <nav
      className={`navbar ${className}`}
      role='navigation'
      aria-label='main navigation'
    >
      {childrenArray.map(
        (c: ReactElement) =>
          c.type === navbarBrandType && React.cloneElement(c, { menuRef })
      )}
      <div ref={menuRef} className={`navbar-menu ${menuClass}`}>
        {childrenArray.filter((c: ReactElement) => c.type === navbarStartType)}
        {childrenArray.filter((c: ReactElement) => c.type === navbarEndType)}
      </div>
    </nav>
  )
}

Navbar.defaultProps = {
  className: '',
  menuClass: ''
}

Navbar.propTypes = {
  className: PropTypes.string,
  menuClass: PropTypes.string,
  children: PropTypes.node
}

export { Navbar, NavbarBrand, NavbarEnd, NavbarStart }
export default Navbar
