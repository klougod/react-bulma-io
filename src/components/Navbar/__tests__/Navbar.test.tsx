import React from 'react'
import Enzyme from 'enzyme'

import Navbar from '../'
import NavbarBrand from '../NavbarBrand'
import NavbarStart from '../NavbarStart'
import NavbarEnd from '../NavbarEnd'

describe('Navbar component test', function () {
  it('should render without crash', function () {
    const heroMock = Enzyme.shallow(<Navbar />)
    expect(heroMock).toMatchSnapshot()
  })

  it('should render without crash with correct children', function () {
    const heroMockProps = {
      className: 'heroClass',
      menuClass: 'heroMenuClass'
    }
    const heroMock = Enzyme.shallow(
      <Navbar {...heroMockProps}>
        <NavbarBrand />
        <NavbarStart />
        <NavbarEnd />
      </Navbar>
    )
    expect(heroMock).toMatchSnapshot()
  })
})
