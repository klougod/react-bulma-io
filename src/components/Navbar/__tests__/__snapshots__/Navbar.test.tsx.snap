// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar component test should render without crash 1`] = `
<nav
  aria-label="main navigation"
  className="navbar "
  role="navigation"
>
  <div
    className="navbar-menu "
  />
</nav>
`;

exports[`Navbar component test should render without crash with correct children 1`] = `
<nav
  aria-label="main navigation"
  className="navbar heroClass"
  role="navigation"
>
  <Component
    className=""
    key=".0"
    menuRef={
      Object {
        "current": null,
      }
    }
  />
  <div
    className="navbar-menu heroMenuClass"
  >
    <Component
      className=""
      key=".1"
    />
    <Component
      className=""
      key=".2"
    />
  </div>
</nav>
`;
